### PR TITLE
C#: Move qualified name computation into `QualifiedName.qll`

### PR DESCRIPTION
**Pull request by hvitved**
This PR moves the logic for computing fully qualified names into `QualifiedName.qll`, and makes it parameterized on how to print unbound generics. This is in preparation for a change to the Models-as-Data format for C#.

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll [hunk 1]
```diff
@@ -1,6 +1,201 @@
...
+      strictconcat(Type t, int i | t = cg.getTypeArgument(i) | getFullName(t), "," order by i)
+  }
+
+  /** Holds if declaration `d` has the qualified name `qualifier`.`name`. */
+  predicate hasQualifiedName(Declaration d, string qualifier, string name) {
+    d =
+      any(ValueOrRefType vort |
+        vort instanceof VoidType and
+        qualifier = "System" and
+        name = "Void"
```
Comment (tamasvajk) [inline csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll:72]: Why do we need to special case `VoidType`? 
Couldn't the below cover it?
```
          not exists(vort.getDeclaringType()) and
          qualifier = vort.getNamespace().getFullName() and
          name = vort.getUndecoratedName()
```

File: csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll [hunk 2]
```diff
@@ -1,6 +1,201 @@
...
+            exists(Type elementType, string name0 |
+              hasQualifiedName(elementType, qualifier, name0) and
+              name = name0 + getDimensionString(at, elementType)
+            )
+          )
+        or
+        hasQualifiedName(vort.(TupleType).getUnderlyingType(), qualifier, name)
+        or
+        qualifier = "System" and
+        name = "Nullable<" + getFullName(vort.(NullableType).getUnderlyingType()) + ">"
```
Comment (tamasvajk) [inline csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll:85]: Is this needed? `NullableType` is a subclass of `ConstructedType`, couldn't the below `any(ConstructedType ct | ...` cover nullable types too?

**Review by tamasvajk — DISMISSED** [review]
LGTM, added some questions.

File: csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll (hunk 1 above)
Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll:72]: Perhaps; I just copied over the existing logic.

File: csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll (hunk 2 above)
Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll:85]: Again, I just copied it from the existing setup, but I can try to remove it.

**Review by michaelnebel — APPROVED** [review]
Looks good to me! Also great to collect all the logic printing the qualified name in one place.

**Review by tamasvajk — APPROVED** [review]

**Pull request merged**